### PR TITLE
(prometheus) fix multiple query bug with table panel

### LINKED ISSUE #9134: [Bug] Inconsistent labels breaks table columns
- What Grafana version are you using?
4.4.1

- What datasource are you using?
Prometheus

- What OS are you running grafana on?
Centos 7.3

- What did you do?
Create a table panel, with a query that returns different labels

- What was the expected result?
Either only common labels get a column, or (preferable) rows where a given column has no value are empty (a customizable default value would be nice)

- What happened instead?
The columns are taken from the labels of the first result. Time series with other labels get their labels mapped to columns in (I guess) returned order.

As an example, using this query: `ALERTS{alertstate="firing"}`, I will get back a number of time series with a wide variety of labels. Here's a few from my installation:

```
ALERTS{alertname="DiskFullWithinThreeDays",alertstate="firing",device="/dev/mapper/vg_data-lv_cassandra",environment="prod",fstype="xfs",instance="some-node",job="node_exporter",mountpoint="/var/lib/cassandra",severity="info"}
ALERTS{alertname="MachineDown",alertstate="firing",environment="stage",instance="some-node",job="wmi_exporter",severity="critical"}
ALERTS{alertname="DestinationUnreachable",alertstate="firing",environment="prod",instance="some-node",job="node_exporter",severity="critical"}
ALERTS{alertname="Cassandra_CompactionPendingTasks",alertstate="firing",environment="prod",instance="some-node",job="cassandra_exporter",severity="warning"}
ALERTS{alertname="BlackboxProbeFail",alertstate="firing",instance="http://some-url",job="blackbox",severity="error",url="http://some-url"}
```

This will render a table with the columns from the first returned series, `alertname,alertstate,device,environment,fstype,instance,job,mountpoint,severity`, and the first row in the table will look fine. The second row, though, does not have these labels, and even the labels that are shared are not in the same order. Here things get weird, the severity label is mapped to the device column, instance to environment, etc.

So we end up with something like this:

alertname|alertstate|device|environment|fstype|instance|job|mountpoint|severity
-|-|-|-|-|-|-|-|-
DiskFullWithinThreeDays|firing|/dev/mapper/vg_data-lv_cassandra|prod|xfs|some-node|node_exporter|/var/lib/cassandra|info
MachineDown|firing|stage|some-node|wmi_exporter|critical
DestinationUnreachable|firing|prod|some-node|node_exporter|critical
Cassandra_CompactionPendingTasks|firing|prod|some-node|cassandra_exporter|warning
BlackboxProbeFail|firing|http://some-url|blackbox|error|http://some-url

Which is quite confusing.

### PR DESCRIPTION
**Pull request by mtanda**
Fix https://github.com/grafana/grafana/issues/9134

### CODE REVIEW

**Comment by mtanda** [comment]
By applying this fix, https://github.com/grafana/grafana/issues/9170 seems work.

**Comment by mtanda** [comment]
add test.

File: public/app/plugins/datasource/prometheus/datasource.ts [hunk 1]
```diff
@@ -130,7 +124,20 @@ export class PrometheusDatasource {
         }
       });
 
+      var tableResponses = _.filter(responseList, (response, index) => {
```
Comment (torkelo) [inline public/app/plugins/datasource/prometheus/datasource.ts]: This code looks a bit messy and feels hacky (the way time series data is handled differently, and table data is merged and then transformed). Wonder if there is a nicer way to process response. Also was there a reason for removing the response.status error handling? Has Prometheus changed how it returns errors?

**Review by torkelo — CHANGES_REQUESTED** [review]
This code looks a bit messy and feels hacky (the way time series data is handled differently, and table data is merged and then transformed). Wonder if there is a nicer way to process response. Also was there a reason for removing the response.status error handling? Has Prometheus changed how it returns errors?

File: public/app/plugins/datasource/prometheus/datasource.ts (hunk 1 above)
Comment (mtanda) [reply public/app/plugins/datasource/prometheus/datasource.ts]: > Also was there a reason for removing the response.status error handling? Has Prometheus changed how it returns errors?
Sorry, it is mistake. revert the change.

Comment (mtanda) [reply public/app/plugins/datasource/prometheus/datasource.ts]: > Wonder if there is a nicer way to process response.

Squashing all time series to only one series, is easiest way.

The table panel doesn't distinguish time series, it convert one datapoint to one table record.
So, I think it is not hacky.
If calling transformMetricDataToTable() for each time series, we need to care about which label key is missing.
find all possible label key before calling transformMetricDataToTable(), and then pass it to transformMetricDataToTable().
You might feel it is messy too...

**Comment by mtanda** [comment]
another approach.

**Comment by mtanda** [comment]
The original issue is fixed by https://github.com/grafana/grafana/pull/10050
Close this.

**Pull request closed without merging**